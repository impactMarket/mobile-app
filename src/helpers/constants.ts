export enum Screens {
    CommunityDetails = 'CommunityDetails',
    FAQ = 'FAQ',
    Welcome = 'Welcome',
    CreateCommunity = 'CreateCommunity',
    WaitingTx = 'WaitingTx',
    Stories = 'Stories',
    NewStory = 'NewStory',
    StoriesCarousel = 'StoriesCarousel',
    Carousel = 'Carousel',
    //
    ClaimExplained = 'ClaimExplained',
    AnonymousReport = 'AnonymousReport',
    AddedBeneficiary = 'AddedBeneficiary',
    RemovedBeneficiary = 'RemovedBeneficiary',
    AddBeneficiary = 'AddBeneficiary',
    AddManager = 'AddManager',
    AddedManager = 'AddedManager',
    //
    Beneficiary = 'Beneficiary',
    WelcomeRulesScreen = 'WelcomeRulesScreen',
    CommunityManager = 'CommunityManager',
    Communities = 'Communities',
    Profile = 'Profile',
    Auth = 'Auth',
}

// async storage
export const STORAGE_USER_ADDRESS = '@wallet:address';
export const STORAGE_USER_PHONE_NUMBER = '@wallet:phonenumber';
export const STORAGE_USER_AUTH_TOKEN = '@user:authtoken';

// redux
// export const INIT_USER = 'INIT_USER';
export const SET_USER_CELO_INFO = 'SET_USER_CELO_INFO';
export const SET_USER_INFO = 'SET_USER_INFO';
export const SET_USER_EXCHANGE_RATE = 'SET_USER_EXCHANGE_RATE';
export const SET_USER_WALLET_BALANCE = 'SET_USER_WALLET_BALANCE';
export const SET_CELO_KIT = 'SET_CELO_KIT';
export const SET_COMMUNITY_CONTRACT = 'SET_COMMUNITY_CONTRACT';
export const SET_COMMUNITY = 'SET_COMMUNITY';
export const SET_USER_IS_BENEFICIARY = 'SET_USER_IS_BENEFICIARY';
export const SET_USER_IS_BLOCKED = 'SET_USER_IS_BLOCKED';
export const SET_USER_IS_SUSPECT = 'SET_USER_IS_SUSPECT';
export const SET_USER_IS_COMMUNITY_MANAGER = 'SET_USER_IS_COMMUNITY_MANAGER';
export const RESET_USER_APP = 'RESET_USER_APP';
export const RESET_NETWORK_APP = 'RESET_NETWORK_APP';
export const SET_PUSH_NOTIFICATION_TOKEN = 'SET_PUSH_NOTIFICATION_TOKEN';
export const SET_AUTH_TOKEN = 'SET_AUTH_TOKEN';
export const SET_USER_LANGUAGE = 'SET_USER_LANGUAGE';
export const SET_EXCHANGE_RATES = 'SET_EXCHANGE_RATES';
export const CONSENT_ANALYTICS = 'CONSENT_ANALYTICS';
export const SHOW_REPORT_CARD = 'SHOW_REPORT_CARD';

export const SET_USER_AUTH_REQUEST = 'SET_USER_AUTH_REQUEST';
export const SET_USER_AUTH_SUCCESS = 'SET_USER_AUTH_SUCCESS';
export const SET_USER_AUTH_FAILURE = 'SET_USER_AUTH_FAILURE';

export const SET_APP_SUSPECT_WRONG_DATETIME = 'app/suspectWrongDateTime';
export const SET_APP_FROM_WELCOME_SCREEN = 'app/fromWelcomeScreen';
export const SET_APP_BENEFICIARY_HAS_ACCEPTED_TERMS =
    'app/hasBeneficiaryAcceptedRulesAlready';
export const SET_APP_MANAGER_HAS_ACCEPTED_TERMS =
    'app/hasManagerAcceptedRulesAlready';

export enum appAction {
    SET_PUSH_NOTIFICATION_LISTENERS = 'app/setPushNotificationListeners',
}

export enum modalDonateAction {
    OPEN = 'donateModal/open',
    GO_TO_CONFIRM_DONATE = 'donateModal/goToConfirm',
    GO_TO_ERROR_DONATE = 'donateModal/goToError',
    GO_BACK_TO_DONATE = 'donateModal/goBackToDonate',
    IN_PROGRESS = 'donateModal/inProgress',
    CLOSE = 'donateModal/close',
}

export enum apiImageTargets {
    COVER = '/community/picture',
    PROFILE = '/user/picture',
}

export enum imageTargets {
    COVER = 'COVER',
    PROFILE = 'PROFILE',
}

export enum storiesAction {
    INIT_REQUEST = 'stories/init_request',
    INIT_SUCCESS = 'stories/init_success',
    INIT_FAILURE = 'stories/init_failure',
    CONCAT = 'stories/concat',
    USER_STORIES_REQUEST = 'stories/user_request',
    USER_STORIES_SUCCESS = 'stories/user_success',
    USER_STORIES_FAILURE = 'stories/user_failure',
}

export const SET_VIEW_MANAGER_DETAILS = 'view/managerDetails';

export enum celoNetwork {
    noAddress = '0x0000000000000000000000000000000000000000',
}
