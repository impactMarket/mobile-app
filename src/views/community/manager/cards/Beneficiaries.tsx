import { useNavigation } from '@react-navigation/native';
import i18n from 'assets/i18n';
import Button from 'components/core/Button';
import Card from 'components/core/Card';
import { Screens } from 'helpers/constants';
import React from 'react';
import { Alert, View } from 'react-native';
import { Headline } from 'react-native-paper';

import SuspiciousActivity from './SuspiciousActivity';
interface IBeneficiariesProps {
    beneficiaries: number | undefined;
    removedBeneficiaries: number | undefined;
    hasFundsToNewBeneficiary: boolean;
    isSuspeciousDetected: boolean | null;
}

function Beneficiaries(props: IBeneficiariesProps) {
    const navigation = useNavigation();

    const {
        beneficiaries,
        removedBeneficiaries,
        hasFundsToNewBeneficiary,
        isSuspeciousDetected,
    } = props;

    return (
        <View>
            <Card>
                <Card.Content>
                    <Headline
                        style={{
                            opacity: 0.48,
                            fontFamily: 'Gelion-Bold',
                            fontSize: 13,
                            fontWeight: '500',
                            lineHeight: 12,
                            letterSpacing: 0.7,
                        }}
                    >
                        {i18n.t('generic.beneficiaries').toUpperCase()}
                    </Headline>
                    <Button
                        modeType="gray"
                        bold
                        disabled={beneficiaries === 0}
                        style={{ marginVertical: 5 }}
                        onPress={() =>
                            navigation.navigate(Screens.AddedBeneficiary)
                        }
                    >
                        {i18n.t('generic.added')} ({beneficiaries})
                    </Button>
                    <Button
                        modeType="gray"
                        bold
                        disabled={removedBeneficiaries === 0}
                        style={{ marginVertical: 5 }}
                        onPress={() =>
                            navigation.navigate(Screens.RemovedBeneficiary)
                        }
                    >
                        {i18n.t('generic.removed')} ({removedBeneficiaries})
                    </Button>
                    <View>
                        {hasFundsToNewBeneficiary ? (
                            <Button
                                modeType="green"
                                bold
                                style={{
                                    marginVertical: 5,
                                }}
                                onPress={() =>
                                    navigation.navigate(Screens.AddBeneficiary)
                                }
                            >
                                {i18n.t('manager.addBeneficiary')}
                            </Button>
                        ) : (
                            <Button
                                modeType="default"
                                icon="alert"
                                style={{
                                    marginVertical: 5,
                                    backgroundColor: '#f0ad4e',
                                }}
                                onPress={() => {
                                    Alert.alert(
                                        i18n.t('manager.noFunds'),
                                        i18n.t(
                                            'manager.notFundsToAddBeneficiary'
                                        ),
                                        [{ text: i18n.t('generic.close') }],
                                        { cancelable: false }
                                    );
                                }}
                            >
                                {i18n.t('manager.addBeneficiary')}
                            </Button>
                        )}
                        {isSuspeciousDetected && <SuspiciousActivity />}
                    </View>
                </Card.Content>
            </Card>
        </View>
    );
}

export default Beneficiaries;
