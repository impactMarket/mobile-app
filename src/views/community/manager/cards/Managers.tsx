import Button from 'components/core/Button';
import Card from 'components/core/Card';
import { IRootState } from 'helpers/types/state';
// import ListManagers from 'components/ListManagers';
import React from 'react';
import { View } from 'react-native';
import { Headline } from 'react-native-paper';
import i18n from 'assets/i18n';
import { useNavigation } from '@react-navigation/native';
import { Screens } from 'helpers/constants';

interface IManagersProps {
    managers: number;
}
function Managers(props: IManagersProps) {
    const navigation = useNavigation();

    const { managers } = props;

    return (
        <View>
            <Card style={{ marginTop: 16 }}>
                <Card.Content>
                    <Headline
                        style={{
                            opacity: 0.48,
                            fontFamily: 'Gelion-Bold',
                            fontSize: 13,
                            fontWeight: '500',
                            lineHeight: 12,
                            letterSpacing: 0.7,
                        }}
                    >
                        {i18n.t('managers').toUpperCase()}
                    </Headline>
                    <Button
                        modeType="gray"
                        bold={true}
                        disabled={managers === 0}
                        style={{ marginVertical: 5 }}
                        onPress={() =>
                            navigation.navigate(Screens.AddedManager)
                        }
                    >
                        {i18n.t('added')} ({managers})
                    </Button>
                    <Button
                        modeType="green"
                        bold={true}
                        style={{
                            marginVertical: 5,
                        }}
                        onPress={() => navigation.navigate(Screens.AddManager)}
                    >
                        {i18n.t('addManager')}
                    </Button>
                </Card.Content>
            </Card>
        </View>
    );
}

export default Managers;
