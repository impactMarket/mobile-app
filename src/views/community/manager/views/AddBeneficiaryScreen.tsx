import { useNavigation } from '@react-navigation/native';
import i18n from 'assets/i18n';
import Button from 'components/core/Button';
import BackSvg from 'components/svg/header/BackSvg';
import { isOutOfTime } from 'helpers/index';
import { setCommunityMetadata } from 'helpers/redux/actions/user';
import { IRootState } from 'helpers/types/state';
import React, { useState } from 'react';
import { Alert, View } from 'react-native';
import { Divider, IconButton, Paragraph, TextInput } from 'react-native-paper';
import { useDispatch, useSelector } from 'react-redux';
import Api from 'services/api';
import { celoWalletRequest } from 'services/celoWallet';
import SuspiciousActivity from '../cards/SuspiciousActivity';

import ScanQR from './ScanQR';

function AddBeneficiaryScreen() {
    const dispatch = useDispatch();
    const navigation = useNavigation();
    const communityContract = useSelector(
        (state: IRootState) => state.user.community.contract
    );
    const communityMetadata = useSelector(
        (state: IRootState) => state.user.community.metadata
    );
    const userAddress = useSelector(
        (state: IRootState) => state.user.wallet.address
    );
    const userBalance = useSelector(
        (state: IRootState) => state.user.wallet.balance
    );
    const kit = useSelector((state: IRootState) => state.app.kit);

    const [inputAddress, setInputAddress] = useState('');
    const [usingCamera, setUsingCamera] = useState(false);
    const [addInProgress, setAddInProgress] = useState(false);

    // TODO: need to be adjusted regarding the API response
    const [isSuspeciousDetected, setIsSuspeciousDetected] = useState(false);

    const handleModalScanQR = async () => {
        let addressToAdd: string;

        if (communityContract === undefined) {
            return;
        }

        if (userBalance.length < 16) {
            Alert.alert(
                i18n.t('failure'),
                i18n.t('notEnoughForTransaction'),
                [{ text: i18n.t('close') }],
                { cancelable: false }
            );
            return;
        }

        try {
            addressToAdd = kit.web3.utils.toChecksumAddress(inputAddress);
        } catch (e) {
            Alert.alert(
                i18n.t('failure'),
                i18n.t('addingInvalidAddress'),
                [{ text: i18n.t('close') }],
                { cancelable: false }
            );
            return;
        }

        const searchResult = await Api.community.searchBeneficiary(
            true,
            addressToAdd
        );
        if (searchResult.length !== 0) {
            Alert.alert(
                i18n.t('failure'),
                i18n.t('alreadyInCommunity'),
                [{ text: i18n.t('close') }],
                { cancelable: false }
            );
            return;
        }

        setAddInProgress(true);
        celoWalletRequest(
            userAddress,
            communityContract.options.address,
            await communityContract.methods.addBeneficiary(addressToAdd),
            'addbeneficiary',
            kit
        )
            .then((tx) => {
                if (tx === undefined) {
                    return;
                }
                // refresh community details
                setTimeout(() => {
                    Api.community
                        .getByPublicId(communityMetadata.publicId)
                        .then((c) => dispatch(setCommunityMetadata(c!)));
                }, 2500);

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('addedNewBeneficiary'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                navigation.goBack();
            })
            .catch(async (e) => {
                let error = 'possibleNetworkIssues';
                if (
                    e.message.includes('nonce') ||
                    e.message.includes('gasprice is less')
                ) {
                    error = 'possiblyValoraNotSynced';
                } else if (e.message.includes('gas required exceeds')) {
                    error = 'unknown';
                    // verify clock time
                    if (await isOutOfTime()) {
                        error = 'clockNotSynced';
                    }
                } else if (e.message.includes('Invalid JSON RPC response:')) {
                    if (
                        e.message.includes('The network connection was lost.')
                    ) {
                        error = 'networkConnectionLost';
                    }
                    error = 'networkIssuesRPC';
                }
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorAddingBeneficiary', { error: i18n.t(error) }),
                    [{ text: i18n.t('close') }],
                    { cancelable: false }
                );
                console.log(typeof JSON.stringify(e), error);
                console.log(e.message, error);
                Api.system.uploadError(
                    userAddress,
                    'add_beneficiary',
                    e,
                    error
                );
            })
            .finally(() => {
                setAddInProgress(false);
            });
    };

    const personalAddressWarningMessageCondition =
        inputAddress.toLowerCase() === userAddress.toLowerCase();

    return (
        <>
            {personalAddressWarningMessageCondition && (
                <View
                    style={{ alignItems: 'center', paddingHorizontal: '20%' }}
                >
                    <IconButton icon="alert" color="#f0ad4e" size={20} />
                    <Paragraph
                        style={{
                            marginHorizontal: 10,
                            textAlign: 'center',
                        }}
                    >
                        {i18n.t('addingYourOwnAddress')}
                    </Paragraph>
                </View>
            )}
            <View style={{ flex: 1, marginHorizontal: 10, marginTop: 20 }}>
                <Divider />
                <View
                    style={{
                        flexDirection: 'row',
                        justifyContent: 'space-between',
                        // width: '100%',
                        paddingHorizontal: 20,
                    }}
                >
                    <TextInput
                        mode="flat"
                        underlineColor="transparent"
                        style={{
                            width: '80%',
                            fontFamily: 'Gelion-Regular',
                            backgroundColor: 'transparent',
                            paddingHorizontal: 0,
                        }}
                        label={i18n.t('beneficiaryAddress')}
                        value={inputAddress}
                        // required
                        onChangeText={(value: string) => setInputAddress(value)}
                    />
                    <IconButton
                        style={{
                            alignSelf: 'center',
                            width: '20%',
                            right: -15,
                        }}
                        icon="qrcode"
                        size={20}
                        onPress={() => setUsingCamera(true)}
                    />
                </View>
                <Divider />
                <Button
                    modeType="green"
                    style={{
                        marginVertical: 10,
                        marginHorizontal: 20,
                    }}
                    disabled={
                        inputAddress.length === 0
                        // (selectButtonInProgress !== undefined &&
                        //     selectButtonInProgress === true)
                    }
                    loading={addInProgress === true}
                    onPress={() => handleModalScanQR()}
                >
                    {i18n.t('addBeneficiary')}
                </Button>
                {isSuspeciousDetected && <SuspiciousActivity />}
            </View>
            <ScanQR
                isVisible={usingCamera}
                onDismiss={() => setUsingCamera(false)}
                callback={(value) => setInputAddress(value)}
            />
        </>
    );
}
AddBeneficiaryScreen.navigationOptions = () => {
    return {
        headerLeft: () => <BackSvg />,
        headerTitle: i18n.t('addBeneficiary'),
    };
};

export default AddBeneficiaryScreen;
