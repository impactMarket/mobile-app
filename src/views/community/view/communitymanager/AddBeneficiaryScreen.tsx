import { useNavigation } from '@react-navigation/native';
import i18n from 'assets/i18n';
import Button from 'components/Button';
import Header from 'components/Header';
import ValidatedTextInput from 'components/ValidatedTextInput';
import { ethers } from 'ethers';
import { iptcColors } from 'helpers/index';
import { IStoreCombinedState, IStoreCombinedActionsTypes } from 'helpers/types';
import React, { useState } from 'react';
import { Alert, View } from 'react-native';
import {
    // Button,
    Divider,
    IconButton,
    Paragraph,
    Portal,
    TextInput,
    Text,
} from 'react-native-paper';
import { useStore } from 'react-redux';
import { celoWalletRequest } from 'services/celoWallet';
import ScanQR from '../../../common/ScanQR';

function AddBeneficiaryScreen() {
    const store = useStore<IStoreCombinedState, IStoreCombinedActionsTypes>();
    const navigation = useNavigation();
    const { user, network, app } = store.getState();

    const [inputAddress, setInputAddress] = useState('');
    const [usingCamera, setUsingCamera] = useState(false);
    const [addInProgress, setAddInProgress] = useState(false);

    const handleModalScanQR = async () => {
        const { communityContract } = network.contracts;
        const { address } = user.celoInfo;
        let addressToAdd: string;

        if (communityContract === undefined) {
            // TODO: do something beatiful, la la la
            return;
        }

        try {
            addressToAdd = ethers.utils.getAddress(inputAddress);
        } catch (e) {
            Alert.alert(
                i18n.t('failure'),
                i18n.t('addingInvalidAddress'),
                [{ text: i18n.t('close') }],
                { cancelable: false }
            );
            return;
        }

        setAddInProgress(true);
        celoWalletRequest(
            address,
            communityContract.options.address,
            await communityContract.methods.addBeneficiary(addressToAdd),
            'addbeneficiary',
            app.kit
        )
            .then(() => {
                setTimeout(
                    () =>
                        // updateCommunityInfo(user.celoInfo.address, props),
                        10000
                );

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('addedNewBeneficiary'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                navigation.goBack();
            })
            .catch(() => {
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorAddingBeneficiary'),
                    [{ text: i18n.t('close') }],
                    { cancelable: false }
                );
            })
            .finally(() => {
                setAddInProgress(false);
            });
    };

    const personalAddressWarningMessageCondition =
        inputAddress.toLowerCase() === user.celoInfo.address.toLowerCase();
    const usedAddressWarningMessageCondition =
        network.community.beneficiaries.added.find(
            (b) => b.address.toLowerCase() === inputAddress.toLowerCase()
        ) !== undefined;

    return (
        <>
            <Header
                title={i18n.t('addBeneficiary')}
                hasBack
                navigation={navigation}
            />
            {personalAddressWarningMessageCondition && (
                <View
                    style={{ alignItems: 'center', paddingHorizontal: '20%' }}
                >
                    <IconButton icon="alert" color="#f0ad4e" size={20} />
                    <Paragraph
                        style={{
                            marginHorizontal: 10,
                            textAlign: 'center',
                        }}
                    >
                        {i18n.t('addingYourOwnAddress')}
                    </Paragraph>
                </View>
            )}
            {usedAddressWarningMessageCondition && (
                <View
                    style={{ alignItems: 'center', paddingHorizontal: '20%' }}
                >
                    <IconButton icon="alert" color="#f0ad4e" size={20} />
                    <Paragraph
                        style={{
                            marginHorizontal: 10,
                            textAlign: 'center',
                        }}
                    >
                        {i18n.t('alreadyInCommunity')}
                    </Paragraph>
                </View>
            )}
            <View style={{ flex: 1, marginHorizontal: 10, marginTop: 20 }}>
                <Divider />
                <View
                    style={{
                        flexDirection: 'row',
                        justifyContent: 'space-between',
                        // width: '100%',
                        paddingHorizontal: 20,
                    }}
                >
                    <TextInput
                        mode="flat"
                        underlineColor="transparent"
                        style={{
                            width: '80%',
                            fontFamily: 'Gelion-Regular',
                            backgroundColor: 'transparent',
                            paddingHorizontal: 0,
                        }}
                        label={i18n.t('beneficiaryAddress')}
                        value={inputAddress}
                        // required
                        onChangeText={(value: string) => setInputAddress(value)}
                    />
                    <IconButton
                        style={{
                            alignSelf: 'center',
                            width: '20%',
                            right: -15,
                        }}
                        icon="qrcode"
                        size={20}
                        onPress={() => setUsingCamera(true)}
                    />
                </View>
                <Divider />
                <Button
                    modeType="green"
                    style={{
                        marginVertical: 10,
                        marginHorizontal: 20,
                    }}
                    disabled={
                        usedAddressWarningMessageCondition ||
                        inputAddress.length === 0
                        // (selectButtonInProgress !== undefined &&
                        //     selectButtonInProgress === true)
                    }
                    loading={addInProgress === true}
                    onPress={() => handleModalScanQR()}
                >
                    {i18n.t('addBeneficiary')}
                </Button>
            </View>
            <ScanQR
                isVisible={usingCamera}
                onDismiss={() => setUsingCamera(false)}
                callback={(value) => setInputAddress(value)}
            />
        </>
    );
}

export default AddBeneficiaryScreen;
