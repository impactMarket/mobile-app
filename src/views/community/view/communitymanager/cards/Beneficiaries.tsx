import { useNavigation } from '@react-navigation/native';
import i18n from 'assets/i18n';
import { ethers } from 'ethers';
import { updateCommunityInfo, iptcColors } from 'helpers/index';
import { IRootState, ICommunityInfo } from 'helpers/types';
import React, { useState, useEffect } from 'react';
import { Alert, View } from 'react-native';
import { Card, Headline, Button } from 'react-native-paper';
import { connect, ConnectedProps } from 'react-redux';
import { celoWalletRequest } from 'services/celoWallet';

import ModalScanQR from '../../../../common/ModalScanQR';
import { BigNumber } from 'bignumber.js';

interface IBeneficiariesProps {
    community: ICommunityInfo;
    updateCommunity: (community: ICommunityInfo) => void;
}
const mapStateToProps = (state: IRootState) => {
    const { user, network, app } = state;
    return { user, network, app };
};
const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type Props = PropsFromRedux & IBeneficiariesProps;

function Beneficiaries(props: Props) {
    const navigation = useNavigation();
    const [openModalAddBeneficiary, setOpenModalAddBeneficiary] = useState(
        false
    );
    const [addInProgress, setAddInProgress] = useState(false);
    const [hasFundsToNewBeneficiary, setHasFundsToNewBeneficiary] = useState(
        true
    );

    // TODO: add here a method to be called when page is refreshed

    useEffect(() => {
        const loadCommunityBalance = async () => {
            if (props.app.kit.contracts !== undefined) {
                const stableToken = await props.app.kit.contracts.getStableToken();
                const cUSDBalanceBig = await stableToken.balanceOf(
                    props.network.contracts.communityContract._address
                );
                // at least five cents
                setHasFundsToNewBeneficiary(
                    new BigNumber(cUSDBalanceBig.toString()).gte(
                        '50000000000000000'
                    )
                );
            }
        };
        loadCommunityBalance();
    }, [props.app.kit]);

    const handleModalScanQR = async (inputAddress: string) => {
        const { user, network } = props;
        const { communityContract } = network.contracts;
        const { address } = user.celoInfo;
        let addressToAdd: string;

        if (communityContract === undefined) {
            // TODO: do something beatiful, la la la
            return;
        }

        try {
            addressToAdd = ethers.utils.getAddress(inputAddress);
        } catch (e) {
            Alert.alert(
                i18n.t('failure'),
                i18n.t('addingInvalidAddress'),
                [{ text: i18n.t('close') }],
                { cancelable: false }
            );
            return;
        }

        setAddInProgress(true);
        celoWalletRequest(
            address,
            communityContract.options.address,
            await communityContract.methods.addBeneficiary(addressToAdd),
            'addbeneficiary',
            props.app.kit
        )
            .then(() => {
                setTimeout(
                    () =>
                        updateCommunityInfo(props.user.celoInfo.address, props),
                    10000
                );

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('addedNewBeneficiary'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                setOpenModalAddBeneficiary(false);
            })
            .catch(() => {
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorAddingBeneficiary'),
                    [{ text: i18n.t('close') }],
                    { cancelable: false }
                );
            })
            .finally(() => {
                setAddInProgress(false);
            });
    };

    return (
        <View>
            <Card elevation={8}>
                <Card.Content>
                    <Headline
                        style={{
                            opacity: 0.48,
                            fontFamily: 'Gelion-Bold',
                            fontSize: 13,
                            fontWeight: '500',
                            fontStyle: 'normal',
                            lineHeight: 12,
                            letterSpacing: 0.7,
                        }}
                    >
                        {i18n.t('beneficiaries').toUpperCase()}
                    </Headline>
                    <Button
                        mode="outlined"
                        disabled={
                            props.community.beneficiaries.added.length === 0
                        }
                        style={{ marginVertical: 5 }}
                        onPress={() =>
                            navigation.navigate('AddedScreen', {
                                beneficiaries:
                                    props.community.beneficiaries.added,
                            })
                        }
                    >
                        {i18n.t('added')} (
                        {props.community.beneficiaries.added.length})
                    </Button>
                    <Button
                        mode="outlined"
                        disabled={
                            props.community.beneficiaries.removed.length === 0
                        }
                        style={{ marginVertical: 5 }}
                        onPress={() =>
                            navigation.navigate('RemovedScreen', {
                                beneficiaries:
                                    props.community.beneficiaries.removed,
                            })
                        }
                    >
                        {i18n.t('removed')} (
                        {props.community.beneficiaries.removed.length})
                    </Button>
                    <View>
                        {hasFundsToNewBeneficiary ? (
                            <Button
                                mode="contained"
                                style={{
                                    marginVertical: 5,
                                    backgroundColor: iptcColors.greenishTeal,
                                    position: 'relative',
                                }}
                                onPress={() => setOpenModalAddBeneficiary(true)}
                            >
                                {i18n.t('addBeneficiary')}
                            </Button>
                        ) : (
                            <Button
                                icon="alert"
                                mode="contained"
                                style={{
                                    marginVertical: 5,
                                    backgroundColor: '#f0ad4e',
                                }}
                                onPress={() => {
                                    Alert.alert(
                                        i18n.t('noFunds'),
                                        i18n.t('notFundsToAddBeneficiary'),
                                        [{ text: i18n.t('close') }],
                                        { cancelable: false }
                                    );
                                }}
                            >
                                {i18n.t('addBeneficiary')}
                            </Button>
                        )}
                    </View>
                </Card.Content>
            </Card>
            <ModalScanQR
                isVisible={openModalAddBeneficiary}
                openInCamera={false}
                onDismiss={() => setOpenModalAddBeneficiary(false)}
                inputText={i18n.t('beneficiaryAddress')}
                selectButtonText={i18n.t('add')}
                selectButtonInProgress={addInProgress}
                callback={handleModalScanQR}
            />
        </View>
    );
}

export default connector(Beneficiaries);
