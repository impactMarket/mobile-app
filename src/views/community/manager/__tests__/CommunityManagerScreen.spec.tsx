import { useNavigation } from '@react-navigation/core';
import { assert } from 'chai';
import { shallow, ShallowWrapper, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Screens } from 'helpers/constants';
import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import Beneficiaries from '../cards/Beneficiaries';
import Managers from '../cards/Managers';
import CommunityManagerScreen from '../index';
import AddBeneficiaryScreen from '../views/AddBeneficiaryScreen';

const mockStore = configureMockStore();
const store = mockStore({});

configure({ adapter: new Adapter() });

describe('CommunityManager test suite', () => {
    const navigate = jest.fn();
    const onChange = jest.fn();
    const handleModalScanQR = jest.fn();

    let screen: ShallowWrapper<any>;
    let card: ShallowWrapper<any>;

    const event = {
        target: {
            value: '0xc1912fEDSKJSA59aE311',
        },
    };

    beforeEach(() => {
        jest.resetAllMocks();
        (useNavigation as jest.Mock).mockReturnValue({ navigate });
    });

    it('should render CommunityManager screen correctly', () => {
        givenScreen();
        thenScreenRenderProperly();
    });

    it('should render Beneficiary card when community is valid', () => {
        givenBeneficiariesCard();
        thenCardRenderProperly();
    });

    it('should render Manager card when community is valid', () => {
        givenManagerCard();
        thenCardRenderProperly();
    });

    it('should navigate to Add Beneficiary screen when addBeneficiary button is pressed', () => {
        givenBeneficiariesCard();
        thenCardRenderProperly();
        whenPressingButton('addBeneficiaryBtn');
        thenAppNavigateToScreen(Screens.AddBeneficiary);
    });

    it('should navigate to Added Beneficiary screen when addedBeneficiary button is pressed', () => {
        givenBeneficiariesCard();
        thenCardRenderProperly();
        whenPressingButton('addedBeneficiaryBtn');
        thenAppNavigateToScreen(Screens.AddedBeneficiary);
    });

    it('should navigate to Removed Beneficiary screen when removedBeneficiary button is pressed', () => {
        givenBeneficiariesCard();
        thenCardRenderProperly();
        whenPressingButton('removedBeneficiaryBtn');
        thenAppNavigateToScreen(Screens.RemovedBeneficiary);
    });

    it('should add Beneficiary when address is valid', () => {
        givenBeneficiariesCard();
        thenCardRenderProperly();
        whenPressingButton('addBeneficiaryBtn');
        thenAppNavigateToScreen(Screens.AddBeneficiary);
        givenAddBeneficiaryScreen();
        whenInformValoraAddressToAdd();
        whenPressingButton('addBeneficiaryBtn');
        thenMethodIsCalled();
    });

    function givenScreen() {
        assert.isDefined(CommunityManagerScreen);

        screen = shallow(
            <Provider store={store}>
                <CommunityManagerScreen />
            </Provider>
        );
    }

    function thenScreenRenderProperly() {
        expect(screen).toMatchSnapshot();
    }

    function thenCardRenderProperly() {
        expect(card).toMatchSnapshot();
    }

    function whenPressingButton(id: string) {
        const btn = card.find(`[testID="${id}"]`);
        expect(btn).toBeDefined();
        btn.simulate('press');
    }

    function thenMethodIsCalled() {
        expect(handleModalScanQR).toHaveBeenCalled();
    }

    function givenManagerCard() {
        assert.isDefined(Managers);

        card = shallow(
            <Provider store={store}>
                <Managers managers={2} />
            </Provider>
        );
    }

    function givenBeneficiariesCard() {
        assert.isDefined(Beneficiaries);

        card = shallow(
            <Beneficiaries
                beneficiaries={2}
                removedBeneficiaries={5}
                hasFundsToNewBeneficiary
                isSuspeciousDetected={false}
            />
        );
    }

    function givenAddBeneficiaryScreen() {
        assert.isDefined(AddBeneficiaryScreen);

        card = shallow(
            <Provider store={store}>
                <AddBeneficiaryScreen />
            </Provider>
        )
            .childAt(0)
            .dive();
    }

    function whenInformValoraAddressToAdd() {
        const textInput = card.find('beneficiaryAddressInput');
        expect(textInput).toBeDefined();
        textInput.simulate('change', event);
        expect(onChange).toHaveBeenCalledWith('0xc1912fEDSKJSA59aE311');
    }

    function thenAppNavigateToScreen(screen: string, params?: any) {
        if (params) {
            expect(navigate).toHaveBeenCalledWith(screen, params);
        } else {
            expect(navigate).toHaveBeenCalledWith(screen);
        }
    }
});
