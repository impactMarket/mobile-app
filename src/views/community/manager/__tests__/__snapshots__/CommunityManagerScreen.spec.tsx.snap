// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunityManager test suite should add Beneficiary when address is valid 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should navigate to Add Beneficiary screen when addBeneficiary button is pressed 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should navigate to Added Beneficiary screen when addedBeneficiary button is pressed 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should navigate to Removed Beneficiary screen when removedBeneficiary button is pressed 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should render Beneficiary card when community is valid 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should render CommunityManager screen correctly 1`] = `ShallowWrapper {}`;

exports[`CommunityManager test suite should render Manager card when community is valid 1`] = `ShallowWrapper {}`;
